app/frontend/wallet/shelley/shelley-transaction-planner.ts

```diff
@@ -1,315 +1,384 @@
+  const totalInput = inputs.reduce((acc, input) => acc + input.coins, 0) + totalRewards
+  const totalInputTokens = aggregateTokens(inputs.map(({tokens}) => tokens))
+  const deposit = computeRequiredDeposit(certificates)
+  const totalOutput = outputs.reduce((acc, {coins}) => acc + coins, 0) + deposit + totalRewards
+  const totalOutputTokens = aggregateTokens(outputs.map(({tokens}) => tokens)).map((token) =>
+    formatToken({...token, quantity: `${token.quantity}`}, -1)
+  )
 
-interface NoTxPlan {
-  estimatedFee: Lovelace
-  error: {code: string}
-}
+  const tokenDifference = aggregateTokens([totalInputTokens, totalOutputTokens]).filter(
+    (token) => token.quantity !== 0
+  )
 
-const checkOutputs = (outputs, fee) => {
-  for (const {coins} of outputs) {
-    if (coins < 1000000 + fee) return false
+  // Cannot construct transaction plan, not enought tokens
+  if (tokenDifference.some(({quantity}) => quantity < 0)) {
+    throw NamedError('CannotConstructTxPlan')
   }
-  return true
-}
 
-export function computeTxPlan(
-  inputs: Array<Input>,
-  outputs: Array<Output>,
-  possibleChange: Output,
-  certs: Array<Cert>,
-  withdrawals: Array<Withdrawal>
-): TxPlan | null {
-  const withdrawalAmount = withdrawals && withdrawals.length ? withdrawals[0].rewards : 0
-  const totalInput = inputs.reduce((acc, input) => acc + input.coins, 0) + withdrawalAmount
-  const deposit = computeRequiredDeposit(certs)
-  const totalOutput =
-    outputs.reduce((acc, output) => acc + output.coins, 0) + deposit + withdrawalAmount
-
-  if (totalOutput > Number.MAX_SAFE_INTEGER) {
+  if (
+    totalOutput > Number.MAX_SAFE_INTEGER ||
+    totalOutputTokens.some(({quantity}) => quantity > Number.MAX_SAFE_INTEGER)
+  ) {
     throw NamedError('CoinAmountError')
   }
 
-  const feeWithoutChange = computeRequiredTxFee(inputs, outputs, certs, withdrawals)
+  const feeWithoutChange = computeRequiredTxFee(inputs, outputs, certificates, withdrawals)
+
   // Cannot construct transaction plan
-  if (totalOutput + feeWithoutChange > totalInput) return null
+  if (totalOutput + feeWithoutChange > totalInput) {
+    throw NamedError('CannotConstructTxPlan')
+  }
 
   // No change necessary, perfect fit
-  if (totalOutput + feeWithoutChange === totalInput) {
-    if (checkOutputs(outputs, 0)) {
-      return {
-        inputs,
-        outputs,
-        change: null,
-        certs,
-        deposit,
-        fee: feeWithoutChange as Lovelace,
-        withdrawals,
-      }
+  if (totalOutput + feeWithoutChange === totalInput && tokenDifference.length === 0) {
+    return {
+      inputs,
+      outputs,
+      change: null,
+      certificates,
+      deposit,
+      fee: feeWithoutChange,
+      withdrawals,
     }
   }
 
+  const change = {
+    ...possibleChange,
+    tokens: tokenDifference,
+  }
+
   const feeWithChange = computeRequiredTxFee(
     inputs,
-    [...outputs, possibleChange],
-    certs,
+    [...outputs, change],
```

Comment:
we should watch out for big amounts of tokens overflowing max uint64 - not a showstopper for the release but should be a part of the overall migration to bigint. Similar issue in cardano-wallet for context: https://github.com/input-output-hk/cardano-wallet/issues/2532